_posts/2018-12-04-imposing-invariants-in-scala.md

```diff
@@ -0,0 +1,359 @@
+
+
+This approach has a disadvantage concerning type erasure which also plagues the next approach using generalized type constraints. We will address the issue there.  
+
+# 3. Using generalized type constrains
+
+This is the most novel of the approaches and, as far as I know, could not be achieved on Java, contrary to the remaining solutions.  
+
+Let us recall the original problem. On  
+
+```scala
+case class Reservation[SeatType <: Seat](
+  userTicket: Ticket[SeatType],
+  guestTicket: Ticket[SeatType]
+)
+```
+the compiler can impose <code>SeatType</code> to have as upper bound the type <code>Seat</code>. It cannot however impose <code>SeatType</code> to be one of the sub-types exclusively. Meaning that upon userTicket: <code>Ticket[VipTribune]</code> and guestTicket: <code>Ticket[NormalSeat]</code>, both <code>VipTribune</code> and <code>NormalSeat</code> can pretend to be type <code>Seat</code> which would result in userTicket: <code>Ticket[Seat]</code> and guestTicket: <code>Ticket[Seat]</code> which in turn conforms with the upper bound constraint (Reservation[SeatType <: Seat]). The issue really is that a given type <code>A</code> is a subtype of itself.
+
+<img class=" size-full wp-image-397 aligncenter" src="https://aerodatablog.files.wordpress.com/2018/12/typehiearchy1.png" alt="ActionBuilder" />      
+
+
+The trick on this approach is to use a generalized type constraint. These are not built-in features of the language (like type bounds) but rather smart ways some smart people have engineered to leverage the type system of Scala to accomplish some more impressive things.
+
+
+There are 2 or 3 generalized type constraints on the standard library of Scala on package **scala.Predef** which is imported by default. One of the most famous is <code><:<</code>.
+
+The one we are going to use is found on the external library <a id="shapeless-ref" href="#shapeless"> Shapeless </a> by Miles Sabin.
+
+It is written <code>=:!=[A, B]</code>; this is a parameterized trait named <code>=:!=</code> with two types. This can also be written <code>A =:!= B</code>. 
+
+I will not try to explain the intricacies of these things. A detailed explanation on related type constraints can be found <a id="union-types-ref" href="#union-types">here</a>. I hope to transmit a feeling on how to use them. 
+
+Each of these type constraints is used with a specific goal. The goal of <code>A =:!= B</code> is to impose that types <code>A</code> and <code>B</code> are different, whatever they might be. This is accomplished via implicits. Specifically, you declare an implicit variable of type <code>A =:!= B</code>, where <code>A</code> and <code>B</code> are some two types you want to make sure are different. I would say that normally, either <code>A</code>, <code>B</code> or both are abstract/parameterized types, like type <code>Ole</code> below.
+
+```scala
+import shapeless.=:!=
+def foo[Ole](a: Ole)(implicit ev: =:!=[Ole, Int]): A = a
```

Comment:
I'd change this to the infix notation: it's more readable and what you generally found on the wild.

Reply:
infix notation would be another source of confusion for people to whom evidences and this kind of stuff is already new.